File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int x, y;
        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    static int N;
    static char[][] arr;
    static boolean[][][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        arr = new char[N][N];
        for(int i = 0; i < N; i++) {
            String S = br.readLine();
            for(int j = 0; j < N; j++) {
                arr[i][j] = S.charAt(j);
            }
        }
        int[] cnt = new int[2];
        visited = new boolean[N][N][2];
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < N; j++) {
                if(!visited[i][j][0]) {
                    BFS(i, j, true);
                    cnt[0]++;
                }
                if(!visited[i][j][1]) {
                    BFS(i, j, false);
                    cnt[1]++;
                }
            }
        }
        System.out.println(cnt[0] + " " + cnt[1]);
    }

    static void BFS(int x, int y, boolean isNormal) {
        Queue<Node> que = new LinkedList<>();
        que.add(new Node(x, y));
        visited[x][y][isNormal ? 0 : 1] = true;
        while(!que.isEmpty()) {
            Node node = que.poll();
            for(int i = 0; i < 4; i++) {
                int nx = node.x + dx[i];
                int ny = node.y + dy[i];
                if(nx < 0 || ny < 0 || nx >= N || ny >= N || visited[nx][ny][isNormal ? 0 : 1]) continue;
                if((arr[node.x][node.y] == arr[nx][ny])
                || (!isNormal && ((arr[node.x][node.y] == 'G' && arr[nx][ny] == 'R')
                || (arr[node.x][node.y] == 'R' && arr[nx][ny] == 'G')))) {
                    que.add(new Node(nx, ny));
                    visited[nx][ny][isNormal ? 0 : 1] = true;
                }
            }
        }
    }
}

File: 백준/Gold/10026. 적록색약/적록색약.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.head >= this.arr.length;
  }
}

let [N, ...map] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim());
N = Number(N);
map = map.map(line => line.split(''));
const visited = Array.from({length : N}, () => Array.from({length : N}, () => new Array(2).fill(false)));
const ans = new Array(2).fill(0);
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
for(let i = 0; i < N; i++) {
  for(let j = 0; j < N; j++) {
    if(!visited[i][j][0]) {
      BFS(i, j, true);
      ans[0]++;
    }
    if(!visited[i][j][1]) {
      BFS(i, j, false);
      ans[1]++;
    }
  }
}
console.log(ans.join(' '));

function BFS(x, y, isNormal) {
  const que = new Queue();
  que.add({x : x, y : y});
  visited[x][y][isNormal ? 0 : 1] = true;
  while(!que.isEmpty()) {
    const node = que.poll();
    for(let i = 0; i < 4; i++) {
      const nx = node.x + dx[i];
      const ny = node.y + dy[i];
      if(nx < 0 || ny < 0 || nx >= N || ny >= N || visited[nx][ny][isNormal ? 0 : 1]) continue;
      if((map[node.x][node.y] === map[nx][ny])
        || (!isNormal && ((map[node.x][node.y] === 'G' && map[nx][ny] === 'R')
      || (map[node.x][node.y] === 'R' && map[nx][ny] === 'G')))) {
        que.add({x : nx, y : ny});
        visited[nx][ny][isNormal ? 0 : 1] = true;
      }
    }
  }
}